README.md
# jsonrpc

FIXME: description

## Installation

Download from http://example.com/FIXME.

## Usage

FIXME: explanation

    $ java -jar jsonrpc-0.1.0-standalone.jar [args]

## Options

FIXME: listing of options this app accepts.

## Examples

...

### Bugs

...

### Any Other Sections
### That You Think
### Might be Useful

## License

Copyright © 2021 FIXME

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2021-04-07
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2021-04-07
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/jsonrpc/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/jsonrpc/compare/0.1.0...0.1.1


src/jsonrpc/test.clj
(ns jsonrpc.test
  (:require
   [clojure.string :as str]

   ;; [clojure.instant :as instant]
   ;; [clojure.spec.alpha :as s]

   ))



(defmacro with-mock
  [[params mapping] & body]


  `(let [server# {}]

     (try
       ~@body
       (finally
         (.close server#))))

  )



(with-mock [{:port 9999 :path "/api"}
            {:get-user-by-id
             {:status 200
              :body {:foo 42}}
             :list-users
             {:status 200
              :body [{:foo 42} {:foo 42}]}

             }
            ]



  )


(with-mock
  {:port 9999
   :path "/api"
   :method->response
   {:get-user-by-id
    {:status 200
     :body {:foo 42}}

    :list-users
    {:status 200
     :body [{:foo 42} {:foo 42}]}

    :delete-user
    {:status 200
     :body [{:foo 42} {:foo 42}]}}}

  (println 42))


test/jsonrpc/core_test.clj
(ns jsonrpc.core-test
  (:require
   [jsonrpc.spec :as spec]
   [jsonrpc.server :as server]

   [clojure.spec.alpha :as s]
   [clojure.test :refer [deftest is]]))


(s/def :math/sum.in
  (s/tuple number? number?))

(s/def :math/sum.out
  number?)


(defn rpc-sum
  [_ a b]
  (+ a b))


(defn user-create
  [_ {:keys [name age email]}]
  {:id 100
   :name name
   :age age
   :email email})


(s/def :user/create.in
  (s/keys :req-un [:user/name
                   :user/age
                   :user/email]))

(s/def :user/create.out
  (s/keys :req-un [:user/id
                   :user/name
                   :user/age
                   :user/email]))


(def config
  {:overrides
   {:math/sum
    {:allow-batch? false}}

   :handlers
   {:math/sum
    {:handler #'rpc-sum
     :spec-in :math/sum.in
     :spec-out :math/sum.out}

    :user/create
    {:handler user-create
     :spec-in :user/create.in
     :spec-out :user/create.out}}})



(deftest test-handler-ok

  (let [rpc {:id 1
             :method :math/sum
             :params [1 2]
             :version "2.0"}
        request {:params rpc}
        handler (server/make-handler config)

        response (handler request)]

    (is (= {:status 200
            :body {:id 1 :jsonrpc "2.0" :result 3}}
           response))))


(deftest test-handler-map-params

  (let [rpc {:id 1
             :method "user/create"
             :params {:name "Ivan"
                      :age 35
                      :email "test@testc.com"}
             :version "2.0"}
        request {:params rpc}
        handler (server/make-handler config)

        response (handler request)]

    (is (= {:status 200
            :body
            {:id 1
             :jsonrpc "2.0"
             :result {:id 100
                      :name "Ivan"
                      :age 35
                      :email "test@testc.com"}}}
           response))))


(s/def :user/create.out-wrong
  (s/keys :req-un [:extra/field
                   :user/id]))


(deftest test-handler-wrong-out-spec

  (let [rpc {:id 1
             :method "user/create"
             :params {:name "Ivan"
                      :age 35
                      :email "test@testc.com"}
             :version "2.0"}
        request {:params rpc}

        config*
        (assoc-in config
                  [:handlers :user/create :spec-out]
                  :user/create.out-wrong)

        handler (server/make-handler config*)

        response (handler request)]

    (is (=

         {:status 500
          :body {:id 1
                 :jsonrpc "2.0"
                 :error {:code -32603
                         :message "Internal error"
                         :data {:method :user/create}}}}

           response))))


(deftest test-handler-notify-ok

  (let [rpc {:method "math/sum"
             :params [1 2]
             :version "2.0"}
        request {:params rpc}
        handler (server/make-handler config)

        response (handler request)]

    (is (= {:status 200 :body nil}
           response))))


(deftest test-handler-wrong-params

  (let [rpc {:id 1
             :method "math/sum"
             :params [1 nil]
             :version "2.0"}
        request {:params rpc}
        handler (server/make-handler config)

        response (handler request)]

    (is (= {:status 400
            :body
            {:id 1
             :jsonrpc "2.0"
             :error {:code -32602
                     :message "Invalid params"
                     :data
                     {:method :math/sum
                      :explain
                      "nil - failed: number? in: [1] at: [1] spec: :math/sum.in\n"}}}}

           response))))


(deftest test-handler-batch-ok

  (let [rpc [{:id 1
              :method "math/sum"
              :params [1 2]
              :version "2.0"}
             {:id 2
              :method "math/sum"
              :params [3 4]
              :version "2.0"}]
        request {:params rpc}
        handler (server/make-handler config)

        response (handler request)]

    (is (= {:status 200
            :body
            [{:id 1 :jsonrpc "2.0" :result 3}
             {:id 2 :jsonrpc "2.0" :result 7}]}
           response))))


(deftest test-handler-batch-one-fails

  (let [rpc [{:id 1
              :method "math/sum"
              :params [1 2]
              :version "2.0"}
             {:id 2
              :method "math/sum"
              :params [3 nil]
              :version "2.0"}
             {:id 3
              :method "math/sum"
              :params [5 6]
              :version "2.0"}]

        request {:params rpc}
        handler (server/make-handler config)

        response (handler request)]

    (is (=

         {:status 200
          :body
          [{:id 1 :jsonrpc "2.0" :result 3}
           {:id 2
            :jsonrpc "2.0"
            :error
            {:code -32602
             :message "Invalid params"
             :data {:method :math/sum
                    :explain "nil - failed: number? in: [1] at: [1] spec: :math/sum.in\n"}}}
           {:id 3 :jsonrpc "2.0" :result 11}]}

         response))))


;; batch exception
;; batch allowed?
;; batch limit
;; batch notification
;; batch parallel
;; method overrides

;; in-validation off
;; out-validation off


src/jsonrpc/app.clj
(ns jsonrpc.app
  (:require

   [clojure.spec.alpha :as s]

   [jsonrpc.server :as server]

   [ring.adapter.jetty :refer [run-jetty]]

   [ring.middleware.params :refer [wrap-params]]
   [ring.middleware.keyword-params :refer [wrap-keyword-params]]
   [ring.middleware.json :refer [wrap-json-params wrap-json-response]]))



(defn add [[a b] _]
  (+ a b))


(s/def ::add.in (s/tuple int? int?))

(s/def ::add.out int?)



(def rpc-methods
  {:math/add
   {:title "Add"
    :description "Sums to numbers"
    :handler add
    :spec-in ::add.in
    :spec-out ::add.out}})


(def rpc-handler
  (server/make-handler {} rpc-methods))


(def app
  (-> rpc-handler
      wrap-keyword-params
      wrap-json-params
      wrap-params
      wrap-json-response))


(defn start-app []
  (run-jetty app {:port 8080 :join? false}))


src/jsonrpc/spec.clj
(ns jsonrpc.spec
  (:require
   [clojure.string :as str]
   [clojure.instant :as instant]
   [clojure.spec.alpha :as s]))


(s/def ::ne-string
  (s/and string? (complement str/blank?)))


(s/def ::id
  (s/or :int int? :string ::ne-string))

(s/def ::version #{"2.0"})

(s/def ::method
  (s/or :keyword keyword? :string ::ne-string))

(s/def ::params-map
  (s/map-of keyword? any?))

(s/def ::params
  (s/or :map ::params-map :vec vector?))

(s/def ::rpc-single
  (s/keys :req-un [::version
                   ::method]
          :opt-un [::id
                   ::params]))

(s/def ::rpc-batch
  (s/coll-of ::rpc-single))

(s/def ::rpc
  (s/or :single ::rpc-single :batch ::rpc-batch))



(defmacro with-conformer
  [[bind] & body]
  `(s/conformer
    (fn [~bind]
      (try
        ~@body
        (catch Exception e#
          ::s/invalid)))))


(s/def ::->int
  (with-conformer [val]
    (cond
      (int? val) val
      (string? val) (Integer/parseInt val)
      :else ::s/invalid)))


(s/def ::->long
  (with-conformer [val]
    (cond
      (int? val) val
      (string? val) (Long/parseLong val)
      :else ::s/invalid)))


(s/def ::->float
  (with-conformer [val]
    (cond
      (float? val) val
      (string? val) (Float/parseFloat val)
      :else ::s/invalid)))


(s/def ::->double
  (with-conformer [val]
    (cond
      (double? val) val
      (string? val) (Double/parseDouble val)
      :else ::s/invalid)))


(s/def ::->date
  (with-conformer [val]
    (cond
      (inst? val) val
      (string? val) (instant/read-instant-date val)
      :else ::s/invalid)))


(s/def ::->bool
  (with-conformer [val]
    (cond
      (boolean? val) val
      (nil? val) false
      (string? val)
      (case (str/lower-case val)
        ("true" "1") true
        ("false" "0" "") true)
      :else ::s/invalid)))


src/jsonrpc/client.clj
(ns jsonrpc.client
  (:require
   [clj-http.client :as client]

   ;; [clojure.string :as str]
   ;; [clojure.spec.alpha :as s]
   ))


(defn get-ns [kw]
  (-> kw namespace keyword))


(def config-defaults
  {:rpc/fn-before-send identity
   :rpc/fn-id :int
   :rpc/multi? false
   :rpc/fn-dispatch-service get-ns
   :rpc/notify? false
   :method :post
   :headers {:user-agent "jsonrpc.client"}
   :socket-timeout 5000
   :connection-timeout 5000
   :throw-exceptions? true
   :as :json
   :content-type :json})


(def config
  {:rpc/method-options {:users/delete {:rpc/notify? true}}

   :url "http://test.com/test"
   :auth ["user" "pass"]})


(def config-multi
  {:rpc/multi? true
   :rpc/services
   {:users
    {:url "http://ccc.com/aaa"
     :auth ["fff" "ggg"]}
    :sales
    {:url "http://bbb.com/ccc"
     :auth ["fff" "ggg"]}}})


(defn generate-id [fn-id]
  (cond
    (= fn-id :int)
    (rand-int 99999)

    (= fn-id :uuid)
    (str (java.util.UUID/randomUUID))

    (fn? fn-id)
    (fn-id)

    :else
    (throw (ex-info "Wrong ID function" {:fn-id fn-id}))))


(defn rpc-inner

  ([config method]
   (rpc-inner config method nil))

  ([config method params]

   (let [config
         (merge config config-defaults)

         {:rpc/keys
          [multi?
           services
           fn-dispatch-service]} config

         config-service
         (when multi?
           (or (some-> method
                       fn-dispatch-service
                       (get services))
               (throw (ex-info "RPC service not found"
                               {:method method}))))

         config-method
         (get-in config [:rpc/method-options method])

         config
         (merge config
                config-service
                config-method)

         {:rpc/keys
          [notify?
           fn-id
           fn-before-send]} config

         id
         (when-not notify?
           (generate-id fn-id))

         payload
         (cond-> {:version "2.0"
                  :method method}
           id
           (assoc :id id)
           params
           (assoc :params params))

         request
         (-> config
             (assoc :form-params payload))]

     (-> request
         fn-before-send
         client/request
         :body))))


(defn call

  ([config method]
   (rpc-inner config method))

  ([config method params]
   (if (map? params)
     (rpc-inner config method params)
     (rpc-inner config method [params])))

  ([config method arg & args]
   (rpc-inner config method (cons arg args))))


(defn with-notify [config method]
  (assoc-in
   config
   [:rpc/method-options method :rpc/notify?]
   true))


(defn notify

  ([config method]
   (call (with-notify config method) method))

  ([config method params]
   (call (with-notify config method) method params))

  ([config method arg & args]
   (apply call (with-notify config method) method arg args)))


(defn batch [client batches]

  )


#_
(batch config
       [:user/get-by-id 1]
       [:user/get-by-id 2]
       [:user/get-by-id 3])



#_
(defn remap-batch
  [[method arg & args]]

  (let [params 1])


  (if (map? arg)
    (if args
      (throw (new Exception "Wrong params"))
      {:params arg})
    {:params (into [arg] args)}))


#_
(defn remap-batches
  [batches]
  (mapv remap-batch batches))



(defn call-batch
  [config batches]

  (rpc-inner config (remap-batches batches))






  )



#_
(rpc config :user/get-by-id 42)

#_
(rpc client :job/poll "34234235235")

#_
(rpc client :users/get-by-id 42)

#_
(rpc-client :users/get-by-id {:id 42})



#_
(rpc-batch client
           [[:users/get-by-id 42]
            [:users/create {:name "Ivan"}]] )

#_
(defn make-multi-client [opt]
  )



#_
(rpc-call config :foo/bar 1 2 3)

#_
(rpc-call config :foo/bar {:name "Ivan"
                           :age 42
                           :email "foo@test.com"})

#_
(rpc-notify config :foo/bar 1 2 3)

#_
(rpc-call config ^:notify [:foo/bar 1 2 3])


#_
(rpc-batch config

            [:foo/bar 1]
           ^:notify [:foo/bar 2]
           [:foo/bar 3]
           )



#_
(def multi-client

  {:dispatch
   (comp keyword namespace)

   :services
   {:users
    {:endpoint "http://test.com/test"
     :auth ["user" "pass"]
     :headers {:foo 42}
     :timeout 30000
     :throw-exceptions? 1}

    :sales
    {:endpoint "http://test.com/test"
     :auth ["user" "pass"]
     :headers {:foo 42}
     :timeout 30000
     :throw-exceptions? 1}}
   }

  {:users 1
   }


  )

#_
(rpc multi-client :users/get-user-by-id 42)


src/jsonrpc/schema.clj
(ns jsonrpc.schema
  (:require
   [clojure.string :as str]
   [clojure.spec.alpha :as s]))


(declare make-schema)


;; (alias 'c 'clojure.core)

(defmulti pred->schema identity)


(defmethod pred->schema :default [_])


(defmacro defpred [symbol schema]
  `(defmethod pred->schema '~symbol [~'_]
     ~schema))


(defpred clojure.core/boolean?
  {:type :boolean})

(defpred clojure.core/string?
  {:type :string})

(defpred clojure.core/int?
  {:type :integer})

(defpred clojure.core/pos-int?
  {:type :integer
   :minimum 0})

(defpred clojure.core/nat-int?
  {:type :integer
   :minimum 1})

(defpred clojure.core/inst?
  {:type :date-time})

(defpred clojure.core/nil?
  {:type :null})

(defpred clojure.core/uuid?
  {:type :string
   :pattern "^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$"})

(defpred clojure.core/float?
  {:type :number})

(defn cons? [x]
  (instance? clojure.lang.Cons x))

(defmulti expr->schema first)

(defmethod expr->schema :default [_])

(defmacro defexpr [dispatch bindings  & body]
  `(defmethod expr->schema ~dispatch
     [~bindings]
     ~@body))


(defexpr `s/keys
  [_ & {:keys [req req-un opt opt-un]}]

  {:type :object
   :properties
   (into {} (for [spec req-un]
              [(name spec) (make-schema spec)]))

   :required
   (for [spec req-un]
     (name spec))})


(defexpr `s/coll-of
  [_ spec-item & {:keys [count
                         distinct
                         min-count
                         max-count]}]

  (cond-> {:type :array
           :items (make-schema spec-item)}

    distinct
    (assoc :uniqueItems true)

    min-count
    (assoc :minItems min-count)

    max-count
    (assoc :maxItems max-count)))



(defexpr `s/tuple
  [_ & specs]
  {:type :array
   :items (mapv make-schema specs)})


(defn form->schema [form]
  (cond
    (symbol? form) (pred->schema form)
    (cons? form) (expr->schema form)))


(defmulti spec->schema identity)


(defmethod spec->schema :default [_] nil)


(defmacro defspec [spec schema]
  `(defmethod spec->schema ~spec [~'_]
     ~schema))


(def default-schema {:type :any})


(defmulti guess-type type)


(defmethod guess-type String [_]
  "string")

(defmethod guess-type Integer [_]
  "integer")

(defmethod guess-type Long [_]
  "integer")


(defn enum->schema [spec]
  (if (= 1 (count spec))
    {:const (first spec)}
    {:enum (into [] spec)}))


(defn make-schema [spec]
  (merge
   default-schema

   (cond

     (set? spec)
     (enum->schema spec)

     (cons? spec)
     (expr->schema spec)

     (qualified-symbol? spec)
     (pred->schema spec)

     (qualified-keyword? spec)
     (merge
      (when-let [form (some-> spec s/get-spec s/form)]
        (make-schema form))
      (spec->schema spec)))))


;;

(s/def ::age int?)

(s/def ::address string?)

(s/def ::addresses (s/coll-of ::address :min-count 2 :max-count 5 :distinct true))

(s/def ::user
  (s/keys :req-un [::age
                   ::addresses
                   ::bbbb]))

(defspec ::age
  {:title "Age"
   :description "User age in years"})


(s/def ::xxx (s/tuple int? int? keyword?))


(s/def ::bbb #{"sdfsd" "ffff" "sdfsdf"})


src/jsonrpc/server.clj
(ns jsonrpc.server
  (:require
   [jsonrpc.spec :as spec]

   [clojure.tools.logging :as log]

   [clojure.spec.alpha :as s]))


;; make name
;; init multiproject

;; check specs in config
;; check functions in config
;; auto-load func ns in config

;; refactor errors
;; better explain (expound?)
;; drop version

;; method overrides

;; spec for config
;; validate config in handler

;; hadnler overrides
;; discover handler
;; rename fucntions


(defn explain-str [spec data]
  (let [out (s/explain-str spec data)]
    (when-not (= out "Success!\n")
      out)))


(def rpc-errors
  {:parse-error
   {:status 500
    :code -32700
    :message "Parse error"}

   :invalid-request
   {:status 400
    :code -32600
    :message "Invalid Request"}

   :not-found
   {:status 404
    :code -32601
    :message "Method not found"}

   :invalid-params
   {:status 400
    :code -32602
    :message "Invalid params"}

   :internal-error
   {:status 500
    :code -32603
    :message "Internal error"}})


(def into-map (partial into {}))


(def code->status
  (into-map
   (map (juxt :code :status)
        (vals rpc-errors))))


(defn rpc-error!
  [params]

  (let [{:keys [type]} params

        rpc-error
        (or (get rpc-errors type)
            (get rpc-errors :internal-error))

        data
        (merge rpc-error params)]

    (throw (ex-info "RPC error" data))))


(defn find-method
  [{:as this :keys [config rpc]}]

  (let [{:keys [id method]} rpc
        {:keys [handlers]} config

        handler-map (get handlers method)]

    (if-not handler-map
      (rpc-error! {:id id
                   :type :not-found
                   :data {:method method}})
      (assoc this
             :handler-map handler-map))))


(defn validate-params
  [{:as this :keys [config rpc handler-map]}]

  (let [{:keys [id method params]}
        rpc

        {:keys [validate-in-spec?]}
        config

        {:keys [spec-in
                spec-out
                handler]} handler-map

        validate?
        (and validate-in-spec? spec-in)

        explain
        (when validate?
          (explain-str spec-in params))]

    (if explain
      (rpc-error! {:id id
                   :type :invalid-params
                   :data {:method method
                          :explain explain}})
      this)))


(defn execute-method
  [{:as this :keys [config request rpc handler-map]}]

  (let [{:keys [params]} rpc
        {:keys [handler]} handler-map

        ;; TODO assert handler (symbol)

        {:keys [pass-request-to-handler?]}
        config

        arg-list
        (if (vector? params)
          params [params])

        arg-list
        (if pass-request-to-handler?
          (cons request arg-list)
          arg-list)

        result
        (apply handler arg-list)]

    (assoc this :result result)))


(defn validate-output
  [{:as this :keys [config result rpc handler-map]}]

  (let [{:keys [spec-out]} handler-map

        {:keys [validate-out-spec?]} config

        {:keys [id method]} rpc

        validate?
        (and validate-out-spec? spec-out)

        explain
        (when validate?
          (explain-str spec-out result))]

    (if explain
      ;; TODO log

      (rpc-error! {:id id
                   :type :internal-error
                   :data {:method method}})

      this)))


(defn compose-response
  [{:as this :keys [rpc result]}]
  (let [{:keys [id version]} rpc]
    (when id
      {:id id
       :jsonrpc version
       :result result})))


(def types-no-log
  #{:invalid-request
    :not-found
    :invalid-params})


(defn rpc-single-error-handler
  [e]
  (let [{:keys [id code message data]}
        (ex-data e)]

    (when-not (contains? types-no-log type)
      (log/error e))

    {:id id
     :jsonrpc "2.0"
     :error {:code code
             :message message
             :data data}}))


(defn coerce-rpc
  [this]
  (update-in this [:rpc :method] keyword))


(defn process-rpc-single
  [this]
  (-> this
      coerce-rpc
      find-method
      validate-params
      execute-method
      validate-output
      compose-response
      (try
        (catch Throwable e
          (rpc-single-error-handler e)))))


(defn guess-http-status
  [{:keys [error]}]
  (if error
    (let [{:keys [code]} error]
      (get code->status code 500))
    200))


(defn check-batch-limit
  [{:as this :keys [config rpc]}]

  (let [{:keys [max-batch-size]} config

        batch-size (count rpc)

        {:keys [id]} rpc

        exeeded?
        (when max-batch-size
          (> batch-size max-batch-size))]

    (if exeeded?
      (rpc-error! {:id id
                   :type :invalid-params
                   :message "Batch size is too large"})
      this)))


(defn execute-batch
  [{:as this :keys [config rpc]}]

  (let [{:keys [parallel-batch?]} config

        fn-map
        (if parallel-batch? pmap map)

        fn-single
        (fn [rpc-single]
          (process-rpc-single
             (assoc this :rpc rpc-single)))]

    (fn-map fn-single rpc)))


(defn process-rpc-batch
  [this]
  (-> this
      check-batch-limit
      execute-batch))


(defn step-1-parse-payload
  [{:as this :keys [config request]}]

  (let [{:keys [data-field]} config

        rpc
        (get request data-field)

        explain
        (explain-str ::spec/rpc rpc)

        batch?
        (when-not explain
          (vector? rpc))]

    (if explain

      (rpc-error! {:type :invalid-request
                   :data {:explain explain}})

      (assoc this
             :rpc rpc
             :batch? batch?))))


(defn step-2-check-batch
  [{:as this :keys [config batch?]}]

  (let [{:keys [allow-batch?]} config]

    (if (and batch? (not allow-batch?))
      (rpc-error! {:type :invalid-request})
      this)))


(defn step-3-process-rpc
  [{:as this :keys [batch?]}]

  (let [result
        (if batch?
          (process-rpc-batch this)
          (process-rpc-single this))]

    (assoc this :result result)))


(defn step-4-http-response
  [{:as this :keys [batch? result]}]

  (if batch?

    (let [result (remove nil? result)]
      {:status 200
       :body result})

    (let [status (guess-http-status result)]
      {:status status
       :body result})))


(def config-default
  {:pass-request-to-handler? true
   :data-field :params
   :validate-in-spec? true
   :validate-out-spec? true
   :allow-batch? true
   :max-batch-size 25
   :parallel-batch? true})


(defn make-handler
  [config]
  (fn [request]

    (-> {:config (merge config-default config)
         :request request}

        step-1-parse-payload
        step-2-check-batch
        step-3-process-rpc
        step-4-http-response

        (try
          (catch Throwable e
            (log/error e)

            (let [{:keys [status code message data]}
                  (ex-data e)]

              {:status (or status 500)
               :body {:error {:code code
                              :message message
                              :data data}}}))))))



;;;;;;;;;



#_
(defn ensure-fn [obj]
  (cond

    (fn? obj)
    obj

    (and (var? obj) (fn? @obj))
    obj

    (and (symbol? obj) (resolve obj))
    (resolve obj)

    :else
    (throw (new Exception (format "Wrong function: %s" obj)))))
